api/src/main/java/org/apache/iceberg/expressions/ResolvedReference.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.expressions;
+
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.types.Types;
+
+public class ResolvedReference<T> extends NamedReference<T> {
+  private final int fieldId;
+
+  public ResolvedReference(String name, int fieldId) {
+    super(name);
+    this.fieldId = fieldId;
+  }
+
+  public int fieldId() {
+    return fieldId;
+  }
+
+  @Override
+  public BoundReference<T> bind(Types.StructType struct, boolean caseSensitive) {
+    Schema schema = struct.asSchema();
+    Types.NestedField field = schema.findField(fieldId);
+    ValidationException.check(
+        field != null,
+        "Cannot find field with id '%s' in struct: %s, since we are resolving based on ID",
+        fieldId,
+        schema.asStruct());
+
+    return new BoundReference<>(field, schema.accessorForField(field.fieldId()), name());
+  }
+
+  @Override
+  public NamedReference<T> ref() {
+    return new NamedReference<>(name());
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    ResolvedReference<?> that = (ResolvedReference<?>) o;
+    return fieldId == that.fieldId && name().equals(that.name());
+  }
+
+  @Override
+  public int hashCode() {
+    return 31 * fieldId + name().hashCode();
+  }
+
+  @Override
+  public String toString() {
+    return String.format("ref(name=\"%s\", fieldId=\"%s\")", name(), fieldId);
```

Comment:
`BoundReference` uses `id` rather than `fieldId`. It also looks like the convention for multi-word names is to use kebab case, as `BoundReference` does for `accessor-type`.